h/api/search/index.py

```diff
@@ -69,3 +73,183 @@ def delete(es, annotation_id):
     except elasticsearch.NotFoundError:
         log.exception('Tried to delete a nonexistent annotation from the '
                       'search index, annotation id: %s', annotation_id)
+
+
+class BatchIndexer(object):
+    """
+    A convenience class for reindexing all annotations from the database to
+    the search index.
+    """
+
+    def __init__(self, session, es_client, request):
+        self.session = session
+        self.es_client = es_client
+        self.request = request
+
+    def index_all(self):
+        """Reindex all annotations, and retry failed indexing operations once."""
+
+        errored = None
+        for _ in range(2):
+            errored = self.index(errored)
+            log.debug(
+                'Failed to index {} annotations, might retry'.format(len(errored)))
+
+            if not errored:
+                break
+
+    def index(self, annotation_ids=None):
+        """
+        Reindex annotations.
+
+        :param annotation_ids: a list of ids to reindex, reindexes all when `None`.
+        :type annotation_ids: collection
+
+        :returns: a set of errored ids
+        :rtype: set
+        """
+        if not annotation_ids:
+            annotations = self._stream_all_annotations()
+        else:
+            annotations = self._stream_filtered_annotations(annotation_ids)
+
+        indexing = es_helpers.streaming_bulk(self.es_client.conn, annotations,
+                                             chunk_size=100,
+                                             raise_on_error=False,
+                                             expand_action_callback=self._present)
+        errored = set()
+        for ok, item in indexing:
+            if not ok:
+                errored.add(item['index']['_id'])
+        return errored
+
+    def _present(self, annotation):
```

Comment:
I'd suggest calling this "prepare" or something a bit more generic -- I'm worried that we're falling slowly into the trap of assuming that we'll always use a presenter to turn an annotation object into a thing for the search index, and that definitely doesn't have to be true!